### PR TITLE
Fix bug: don't remove search params after page load

### LINKED ISSUE #163: Bug: On navigating to the homepage with search params, the search params should not disappear
# Current situation
If you open a URL to the homepage with search params, the search params disappear. With this behaviour, it is not possible to share a filter.

# Wanted situation
If you open a URL to the homepage with search params, the search params stay and filter the datasets with the filters in the search params.

### PR DESCRIPTION
**Pull request by barbarah**
This will fix #163 

The search params were not set at all in the client filters.

I also renamed `sort` to `sortBy`. When using `useSearchParams`, the key `sort` is not allowed. Eventually, I did not use `useSearchParams` (client-side function) but went the server-side route. To keep it future-proof, I still kept the rename.

### CODE REVIEW

File: src/lib/search-params.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+// Next.js already separate the search query string into separate properties with string values.
+// Always return a valid SearchOptions object, even if the search params aren't correct,
+// so the application doesn't fail on invalid search params.
+export function fromSearchParamsToSearchOptions({
+  publishers,
+  licenses,
+  query,
+  offset = '0',
+  sortBy = defaultSortBy,
+}: FromSearchParamsToSearchOptionsProps): Response {
```
Comment (sdevalk) [inline src/lib/search-params.ts:100]: Consideration: you could add a (Zod) validator here, for validating the values, setting default values and transforming the values (e.g. from strings to other types)?

File: src/lib/search-params.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+  let sortByResponse = sortBy;
```
Comment (sdevalk) [inline src/lib/search-params.ts:101]: Would it be possible to use a bit more descriptive name for this variable? It's the only one that uses the suffix 'Response', and it's not immediately clear what the meaning of 'Response' is.

Comment (sdevalk) [reply src/lib/search-params.ts:101]: Also, I do not instantly understand the changes that are made to the variable's value. Initially it's the default value (line 99), then it's also the default value (line 119)?

File: src/lib/search-params.ts (hunk 1 above)
Comment (sdevalk) [inline src/lib/search-params.ts:100]: The return type is `Response`, but the function name suggests it transforms search params to search options. Would it be an idea to name the return type `SearchOptions`?

File: src/lib/search-params.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+  publishers?: string;
+  licenses?: string;
+  query?: string;
+  offset?: string;
+  sortBy?: SortBy;
+}
+
+interface Response {
+  searchOptions: SearchOptions;
+  sortBy: SortBy;
```
Comment (sdevalk) [inline src/lib/search-params.ts:87]: Why is `sortBy` a separate property (and not a part of the search options)? (Perhaps add a comment?)

File: src/lib/search-params.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    sortBy: SortBySearchOption.Name,
+    sortOrder: SortOrder.Ascending,
+  },
+  [SortBy.NameDesc]: {
+    sortBy: SortBySearchOption.Name,
+    sortOrder: SortOrder.Descending,
+  },
+};
+
+interface ToSearchParamsProps {
```
Comment (sdevalk) [inline src/lib/search-params.ts:30]: You could omit the `To` here? The word only has meaning in a specific context (if you know a transformation occurred from something to something else), not outside of it?

Also, `ParamsProps` sounds a bit redundant :) Perhaps `SearchParams` could suffice?

File: src/app/[locale]/page.tsx [hunk 5]
```diff
@@ -2,97 +2,23 @@ import datasetFetcher from '@/lib/dataset-fetcher-instance';
...
 
+  let hasError, searchResult;
   try {
-    const searchResult = await datasetFetcher.search(validOptions);
-    return {searchResult, hasError: false};
-  } catch (error) {
-    return {hasError: true, errorMessage: 'Fetch error', searchResult: null};
+    searchResult = await datasetFetcher.search(searchOptions);
+  } catch {
+    hasError = true;
```
Comment (sdevalk) [inline src/app/[locale]/page.tsx:20]: Should we log the error? We're not using a logger at the moment (I need one soon in the backend code :)) but a `console.error` could be sufficient?

File: src/lib/search-params.ts [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+  sortBy?: SortBy;
+}
+
+interface Response {
+  searchOptions: SearchOptions;
+  sortBy: SortBy;
+}
+
+// This function translates the search params to valid search options.
+// Next.js already separate the search query string into separate properties with string values.
```
Comment (sdevalk) [inline src/lib/search-params.ts:91]: Minor: `separates`?

**Review by sdevalk — COMMENTED** [review]
Cool! Also nice work on the tests!

File: src/lib/search-params.ts (hunk 3 above)
Comment (barbarah) [reply src/lib/search-params.ts:87]: This is the sortBy used by the client. This is a combination of sortBy and sortOrder in the searchOptions. Let me see how I can make this more clear

Comment (barbarah) [reply src/lib/search-params.ts:87]: I decided to remove this return value and only return searchOptions. This just makes the whole function confusing. The client can figure out later what value it needs. That this function can just do one thing.

File: src/lib/search-params.ts (hunk 2 above)
Comment (barbarah) [reply src/lib/search-params.ts:101]: I have removed this

File: src/lib/search-params.ts [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+      publishers: publishers?.split(',').filter(id => !!id),
+      licenses: licenses?.split(',').filter(id => !!id),
+    },
+    sortBy: sortBySearchOption,
+    sortOrder: sortOrder,
+    query: query,
+  };
+
+  const validOptions: SearchOptionsWithRequiredSort =
+    searchOptionsWithFallbackSchema.parse(options);
```
Comment (sdevalk) [inline src/lib/search-params.ts:135]: Nice! A thought (not required for now!): would it be possible to make the transformation in `const options` also a part of the Zod schema (e.g. converting the offset string to a number, splitting the publishers string into an array)? Then we'd have all validation and transformation in one place?

File: src/lib/search-params.ts [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+
+  return validOptions;
+}
+
+interface sortPairProps {
```
Comment (sdevalk) [inline src/lib/search-params.ts:140]: Minor: capital `S` for an interface, `SortPairProps`?

**Review by sdevalk — COMMENTED** [review]
Nice improvements! I've added some minor questions/suggestions.

**Review by sdevalk — APPROVED** [review]

**Pull request merged**